### PR TITLE
*: run BackupRanges in parallel, change meaning of backup concurrency

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Step 2 of #394.

### What is changed and how it works?

Previously `BackupRanges` execute each `BackupRange` sequentially. When the number of ranges is lot and each range cannot spawn too many backup jobs, this will create a bottleneck.

This PR changes `BackupRanges` to execute each `BackupRange` in parallel. 

As pointed out in #394, the meaning of "concurrency" is changed to mean the number of `BackupRange`s being executed in parallel, rather than the thread pool size in TiKV (this also prevents killing the TiKV servers by supplying an extremely large concurrency).

`BackupRawRanges` is currently not changed since it only has a single range.

**Update 2020/07/04**: It was found that running `BackupRanges` in parallel will cause a race condition since all writes to the same internal field `bc.backupMeta.Files`, and causes some records to be lost. Thus we refactored `backup.Client` to remove the `backupMeta` field entirely to prevent any write-write conflicts. The `backupMeta` object is constructed using the returned arrays from functions with explicit coordination between goroutines.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

 - Possible performance regression

Related changes

### Release Note

 - Improve parallel processing power when backing up a large database with many regions.

### CODE REVIEW

File: pkg/task/backup.go [hunk 1]
```diff
@@ -188,10 +188,9 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
 		StartVersion: cfg.LastBackupTS,
 		EndVersion:   backupTS,
 		RateLimit:    cfg.RateLimit,
-		Concurrency:  cfg.Concurrency,
+		Concurrency:  defaultBackupConcurrency,
```
Comment (3pointer) [inline pkg/task/backup.go:191]: If we only need to backup a few ranges in one backup task, for example backup one table without index,  we can not control the `defaultBackupConcurrency` to a proper value.

Comment (kennytm) [reply pkg/task/backup.go:191]: yes that's why we also need the "Step 1" of #394 to complete the picture.

Comment (3pointer) [reply pkg/task/backup.go:191]: got it

**Comment by kennytm** [comment]
There is potential chance this PR causing data loss (detected as checksum mismatch). Please don't merge yet.

**Comment by 3pointer** [comment]
These backupmeta [changes](https://github.com/pingcap/br/blob/d3ec9f4cbd0777735e17d2f9418dcedd6201680a/pkg/backup/client.go#L516-L536) caused data loss,  we should add lock for it.

**Comment by overvenus** [comment]
Please add a check that requires TiKV > v4.0.2 .

**Comment by kennytm** [comment]
@overvenus The `req.concurrency` is still passed downstream. For ≤4.0.2 it will just run using concurrency = 4.

**Pull request merged**